File: example-java/app/src/main/java/com/vibes/androidsdkexampleapp/fragments/VibesMainFragment.java
package com.vibes.androidsdkexampleapp.fragments;

import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;
import android.os.Bundle;
import androidx.fragment.app.Fragment;
import androidx.core.content.ContextCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.vibes.androidsdkexampleapp.R;
import com.vibes.androidsdkexampleapp.api.VibesAPI;
import com.vibes.androidsdkexampleapp.model.SharedPrefsManager;
import com.vibes.androidsdkexampleapp.modelViews.VibesViewModel;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * A simple {@link Fragment} subclass.
 * Activities that contain this fragment must implement the
 * to handle interaction events.
 * Use the {@link VibesMainFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class VibesMainFragment extends Fragment {
    @BindView(R.id.deviceIdView) TextView deviceIdView;
    @BindView(R.id.authTokenView) TextView authTokenView;
    @BindView(R.id.registeredLabelView) TextView registeredLabelView;
    @BindView(R.id.deviceRegBtn) Button deviceRegBtn;
    @BindView(R.id.pushRegBtn) Button pushRegBtn;
    @BindView(R.id.loadingBar) ProgressBar loadingBar;
    private VibesViewModel vibesVM;

    public VibesMainFragment() { }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     * @return A new instance of fragment VibesMainFragment.
     */
    public static VibesMainFragment newInstance() {
        VibesMainFragment fragment = new VibesMainFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_vibes_main, container, false);
        ButterKnife.bind(this, view);
        vibesVM = new ViewModelProvider(this).get(VibesViewModel.class);
        vibesVM.setAPI(new VibesAPI(getContext()));
        vibesVM.setSharedPrefs(SharedPrefsManager.getInstance(getContext()));
        setupSubscribers();
        return view;
    }

    private void setupSubscribers() {
        final Observer<Boolean> observerDisplayLoadingBar = isVisible ->
                loadingBar.setVisibility(isVisible ? View.VISIBLE : View.GONE);
        final Observer<String> observerDeviceIdName = value -> deviceIdView.setText(value);
        final Observer<String> observerAuthToken = token -> authTokenView.setText(token);
        final Observer<Integer> observerDeviceRegName = value -> deviceRegBtn.setText(value);
        final Observer<Boolean> observerPusRegEnabled = value -> pushRegBtn.setEnabled(value);
        final Observer<Integer> observerPushRegColor = color ->
                pushRegBtn.setBackgroundColor(ContextCompat.getColor(getContext(), color));
        final Observer<Integer> observerPushRegText = text -> pushRegBtn.setText(text);
        final Observer<Integer> observerPushRegLabelColor = color ->
                registeredLabelView.setBackgroundColor(ContextCompat.getColor(getContext(), color));
        final Observer<Integer> observerRegisterLabel = text -> registeredLabelView.setText(text);

        vibesVM.getDisplayLoadingBarVisible().observe(getViewLifecycleOwner(), observerDisplayLoadingBar);
        vibesVM.getDeviceIDLabelValue().observe(getViewLifecycleOwner(), observerDeviceIdName);
        vibesVM.getAuthTokenLabelValue().observe(getViewLifecycleOwner(), observerAuthToken);
        vibesVM.getDeviceRegButtonName().observe(getViewLifecycleOwner(), observerDeviceRegName);
        vibesVM.getPushRegButtonEnabled().observe(getViewLifecycleOwner(), observerPusRegEnabled);
        vibesVM.getPushRegButtonColor().observe(getViewLifecycleOwner(), observerPushRegColor);
        vibesVM.getPushRegButtonName().observe(getViewLifecycleOwner(), observerPushRegText);
        vibesVM.getPushRegLabelColor().observe(getViewLifecycleOwner(), observerPushRegLabelColor);
        vibesVM.getPushRegLabelValue().observe(getViewLifecycleOwner(), observerRegisterLabel);
    }

    @OnClick(R.id.deviceRegBtn)
    public void RegDeviceClicked(View view) {
        vibesVM.registerDeviceButtonClicked();
    }

    @OnClick(R.id.pushRegBtn)
    public void RegPushClicked(View view) {
        vibesVM.registerPushButtonClicked();
    }
}

File: example-kotlin/app/src/main/java/com/android/exampleapp/kotlin/fragments/VibesMainFragment.kt
package com.android.exampleapp.kotlin.fragments

import androidx.lifecycle.Observer
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.core.content.ContextCompat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import androidx.lifecycle.ViewModelProvider
import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import com.android.exampleapp.kotlin.R
import com.android.exampleapp.kotlin.api.VibesAPI
import com.android.exampleapp.kotlin.modelViews.VibesViewModel
import com.android.exampleapp.kotlin.utils.SharedPrefsManager


/**
 * A simple [Fragment] subclass.
 * Activities that contain this fragment must implement the
 * [VibesMainFragment.OnFragmentInteractionListener] interface
 * to handle interaction events.
 * Use the [VibesMainFragment.newInstance] factory method to
 * create an instance of this fragment.
 *
 */
class VibesMainFragment : androidx.fragment.app.Fragment() {
    @BindView(R.id.deviceIdView)
    @JvmField var deviceIdView: TextView? = null
    @BindView(R.id.authTokenView)
    @JvmField var authTokenView: TextView? = null
    @BindView(R.id.registeredLabelView)
    @JvmField var registeredLabelView: TextView? = null
    @BindView(R.id.deviceRegBtn)
    @JvmField var deviceRegBtn: Button? = null
    @BindView(R.id.pushRegBtn)
    @JvmField var pushRegBtn: Button? = null
    @BindView(R.id.loadingBar)
    @JvmField var loadingBar: ProgressBar? = null
    private var vibesVM: VibesViewModel? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.vibes_main_fragment, container, false)
        ButterKnife.bind(this, view)
        vibesVM = ViewModelProvider(this).get(VibesViewModel::class.java)
        vibesVM!!.setAPI(VibesAPI(context))
        vibesVM!!.setSharedPrefs(SharedPrefsManager.getInstance(context))
        setupSubscribers()
        return view
    }

    private fun setupSubscribers() {
        val observerDisplayLoadingBar = Observer { isVisible: Boolean? -> loadingBar!!.visibility = if (isVisible!!) View.VISIBLE else View.GONE }
        val observerDeviceIdName = Observer { value: String? -> deviceIdView!!.text = value }
        val observerAuthToken = Observer { token: String? -> authTokenView!!.text = token }
        val observerDeviceRegName = Observer { value: Int? -> deviceRegBtn!!.setText(value!!) }
        val observerPusRegEnabled = Observer { value: Boolean? -> pushRegBtn!!.isEnabled = value!! }
        val observerPushRegColor = Observer { color: Int? -> pushRegBtn!!.setBackgroundColor(ContextCompat.getColor(requireContext(), color!!)) }
        val observerPushRegText = Observer { text: Int? -> pushRegBtn!!.setText(text!!) }
        val observerPushRegLabelColor = Observer { color: Int? -> registeredLabelView!!.setBackgroundColor(ContextCompat.getColor(requireContext(), color!!)) }
        val observerRegisterLabel = Observer { text: Int? -> registeredLabelView!!.setText(text!!) }
        vibesVM!!.getDisplayLoadingBarVisible().observe(viewLifecycleOwner, observerDisplayLoadingBar)
        vibesVM!!.getDeviceIDLabelValue().observe(viewLifecycleOwner, observerDeviceIdName)
        vibesVM!!.getAuthTokenLabelValue().observe(viewLifecycleOwner, observerAuthToken)
        vibesVM!!.getDeviceRegButtonName().observe(viewLifecycleOwner, observerDeviceRegName)
        vibesVM!!.getPushRegButtonEnabled().observe(viewLifecycleOwner, observerPusRegEnabled)
        vibesVM!!.getPushRegButtonColor().observe(viewLifecycleOwner, observerPushRegColor)
        vibesVM!!.getPushRegButtonName().observe(viewLifecycleOwner, observerPushRegText)
        vibesVM!!.getPushRegLabelColor().observe(viewLifecycleOwner, observerPushRegLabelColor)
        vibesVM!!.getPushRegLabelValue().observe(viewLifecycleOwner, observerRegisterLabel)
    }
    companion object {
        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @param param1 Parameter 1.
         * @param param2 Parameter 2.
         * @return A new instance of fragment VibesMainFragment.
         */
        // TODO: Rename and change types and number of parameters
        @JvmStatic
        fun newInstance() =
                VibesMainFragment().apply {
                    arguments = Bundle().apply {
                    }
                }
    }

    @OnClick(R.id.deviceRegBtn)
    fun RegDeviceClicked(view: View?) {
        vibesVM!!.registerDeviceButtonClicked()
    }

    @OnClick(R.id.pushRegBtn)
    fun RegPushClicked(view: View?) {
        vibesVM!!.registerPushButtonClicked()
    }
}
